packages/accounts/src/msca/plugins/session-key/extension.ts

```diff
@@ -104,11 +111,84 @@ export const sessionKeyPluginActions: <
     addSessionKey,
     rotateSessionKey,
     updateKeyPermissions,
+    executeWithSessionKey,
     ...og
   } = sessionKeyPluginActions_<TTransport, TChain, TAccount>(client);
 
+  // The session key plugin returns an opaque error when the permissions check
+  // fails. We want the SDK to return a helpful error instead. After validation
+  // fails, rerun the user op simulation but using state overrides to replace
+  // the code of the session key plugin with a version that produces descriptive
+  // errors.
+  const fixedExecuteWithSessionKey: typeof executeWithSessionKey = async (
+    ...originalArgs
+  ) => {
+    let initialError: unknown;
+    try {
+      return await executeWithSessionKey(...originalArgs);
+    } catch (error) {
+      initialError = error;
+    }
+    const details = getRpcErrorMessageFromViemError(initialError);
+    if (!details?.includes("AA23 reverted (or OOG)")) {
+      throw initialError;
+    }
+    if (!isSmartAccountClient(client) || !client.chain) {
+      throw initialError;
+    }
+    const {
+      args,
+      overrides,
+      context,
+      account = client.account,
+    } = originalArgs[0];
+    if (!account) {
+      throw initialError;
+    }
+    const data = og.encodeExecuteWithSessionKey({ args });
+    const sessionKeyPluginAddress =
+      SessionKeyPlugin.meta.addresses[client.chain.id];
+    // The bytecode is 10kb of hex. Only import it if needed.
+    const { DEBUG_SESSION_KEY_BYTECODE } = await import(
+      "./debug-session-key-bytecode.js"
+    );
+    const updatedOverrides: UserOperationOverrides<TEntryPointVersion> = {
+      ...(overrides as UserOperationOverrides<TEntryPointVersion>),
+      stateOverride: [
+        ...(overrides?.stateOverride ?? []),
+        {
+          address: sessionKeyPluginAddress,
+          code: DEBUG_SESSION_KEY_BYTECODE,
+        },
+      ],
```

Comment:
This use of state overrides is currently broken and depends on the fix in https://github.com/alchemyplatform/aa-sdk/pull/720.